Sprint 05

Yuuki-that-one opened the pull request:
выполненное задание 5 спринта

Commit:
выполненное задание 5 спринта

Commit:
выполненное задание 5 спринта

helecta commented on MovieQuiz/Models/QuizQuestion.swift:
```diff
@@ -0,0 +1,17 @@
+
+struct QuizQuestion {
+    // строка с названием фильма,
+    // совпадает с названием картинки афиши фильма в Assets
```
Здорово, что добавляешь комментарии. Если добавлять через /// , а не // , то тогда можно будет быстро прочитать документацию, если зажать кнопку option и навести на класс/свойство/метод.

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -47,93 +19,110 @@ final class MovieQuizViewController: UIViewController {
+        
+        // Отключение кнопок
+        yesButton.isEnabled = false
+        noButton.isEnabled = false
```
Удобнее было бы вынести изменение состояния кнопок в отдельный метод: 
`private func setButtonsEnabled(_ isEnabled : Bool)`

helecta commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -47,93 +19,110 @@ final class MovieQuizViewController: UIViewController {
+        //Добавление действия к алерту
+        alert.addAction(action)
+        //Показ алерта
+        self.present(alert, animated: true, completion: nil)
```
Тут необходимо использовать AlertPresenter. 
AlertPresenter - отвечает за показ алерта, он необходим чтобы убрать эту задачу с MovieQuizViewController

helecta commented on MovieQuiz/Services/AlertProtocol.swift:
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+protocol AlertPresenterDelegate: AnyObject {
+    func showAlert(quiz result: AlertModel)
```
> Обработка нажатия на кнопку алерт реализована с использованием либо замыкания, либо делегата.

Тут имелся в виду не делегат, который занимается показом алерта, а делегат, который отвечает за нажатие на кнопку в алерте.
 
Либо так через замыкание:
```
struct AlertModel {
    let title: String
    let text: String
    let buttonText: String
    var completion: ((UIAlertAction) -> Void)?

```

Либо так через делегат:
```
protocol AlertPresenterDelegate: AnyObject {
    func alertButtonTapped()
}

class AlertPresenter {
    
    weak var delegate: AlertPresenterDelegate?
    
    func showAlert(controller: UIViewController, alertModel: AlertModel) {
        
        let alert = UIAlertController(title: alertModel.title, message: alertModel.text, preferredStyle: .alert)
        
        let action = UIAlertAction(title: alertModel.buttonText, style: .default) { [weak self] _ in
            self?.delegate?.alertButtonTapped()
        }
        
        alert.addAction(action)
        controller.present(alert, animated: true, completion: nil)
    }
}
```

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,134 @@
+    var totalAccuracy: Double {
+        get {
+            guard let data = userDefaults.data(forKey: Keys.total.rawValue),
+                  let totalAcc = try? JSONDecoder().decode(Double.self, from: data) else {
```
UserDefaults поддерживает работу с примитивными типами [для чтения](https://developer.apple.com/documentation/foundation/userdefaults/1407405-integer) и [записи](https://developer.apple.com/documentation/foundation/userdefaults/1413614-set).
Совсем не обязательно простой тип гонять через декодер, когда есть простые способы данные получить/записать.

helecta commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,134 @@
+        totalAmount += amount
+        
+        // Расчет точности исходя из истории игр
+        totalAccuracy = Double(totalCorrectAnswers) / Double(totalAmount) * 100
```
Лучше тут добавить проверку что totalAmount не равен 0, иначе это потенциальное место где приложение может упасть( деление на 0).

Commit:
исправление замечаний
1) Обработка отображения алерта вынесена в AlertPresenter
2) Переработаны get set статистики
3) добавлена проверка деления на 0 при вычислении точности
4) блокировка кнопок вынесена в отдельную функцию  
